### PR TITLE
sys/psa_crypto: sha3 support

### PR DESCRIPTION
**Pull request by Wunderbaeumchen99817**
**Contribution description**
This PR adds a new pseudomodule for using the existing sha3-hashings-algorithms in a simpler way. 
More specific: It adds gluecode for the PSA Crypto API to access the RIOT implementation of SHA3.

**Testing procedure**
The corresponding tests are located in tests/sys/psa_crypto_hashes/example_sha3_glue.c

You can run them the following way:
1. go to tests/sys/psa_crypto_hashes/
2. run "make compile-commands"
3. run "make flash test"

**Issues/PRs references**
N/A

### CODE REVIEW

File: tests/sys/psa_crypto_hashes/sha3.c [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+};
+
+/**
+ * @brief   Compares the computed hashes with the expected hashes. 
+ * Hashes are computed using the PSA algorithm dispatch.
+ *
+ * @return  psa_status_t
+ */
+psa_status_t sha_3_glue_test(void) {
+    // do sha3-256
```
Comment (Teufelchen1) [inline tests/sys/psa_crypto_hashes/sha3.c:25]:
```suggestion

```

File: sys/include/hashes/psa/riot_hashes.h [hunk 2]
```diff
@@ -77,6 +77,13 @@ typedef sha512_256_context_t psa_hashes_sha512_256_ctx_t;
 #include "hashes/sha256.h"
 #endif
 
+#if (IS_USED(MODULE_PSA_RIOT_HASHES_SHA_3))
+#include "hashes/sha3.h"
+
+typedef keccak_state_t psa_hashes_sha3_ctx_t;
```
Comment (Teufelchen1) [inline sys/include/hashes/psa/riot_hashes.h:83]: Is there a specific reason for that name?

**Comment by Teufelchen1** [comment]
Cool! Could you clean up your commits?

File: sys/include/hashes/psa/riot_hashes.h (hunk 2 above)
Comment (mguetschow) [reply sys/include/hashes/psa/riot_hashes.h:83]: Do you mean `keccak_state_t`? That comes from `hashes/sha3.h`.

File: sys/hashes/psa_riot_hashes/sha_3.c [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+#include "psa/crypto.h"
+#include "hashes/psa/riot_hashes.h"
+
+psa_status_t psa_hashes_sha3_update(psa_hashes_sha3_ctx_t *ctx,
+                                      const uint8_t *input,
+                                      size_t input_length)
+{
+    sha3_update((keccak_state_t *)ctx, input, input_length);
+    return PSA_SUCCESS;
```
Comment (mguetschow) [inline sys/hashes/psa_riot_hashes/sha_3.c:28]: Unfortunate that all of those functions don't return an error code, but it is what it is :)

File: sys/include/hashes/psa/riot_hashes.h [hunk 4]
```diff
@@ -77,6 +77,13 @@ typedef sha512_256_context_t psa_hashes_sha512_256_ctx_t;
...
+#include "../psa_crypto/include/psa_hashes.h"
```
Comment (mguetschow) [inline sys/include/hashes/psa/riot_hashes.h:84]: Why is this needed for SHA-3 and not for the other hash software implementations?

File: sys/psa_crypto/include/psa_hashes.h [hunk 5]
```diff
@@ -341,6 +341,82 @@ psa_status_t psa_hashes_sha512_256_finish(psa_hashes_sha512_256_ctx_t *ctx,
...
 
+#if IS_USED(MODULE_PSA_HASH_SHA_3) || defined(DOXYGEN)
+/**
+ * @brief   Low level wrapper function to call a driver for a SHA3-256 hash setup
+ *          See @ref psa_hash_setup()
+ *
+ * @param ctx
+ * @return psa_status_t
+ */
+psa_status_t psa_hashes_sha3_256_setup(psa_hashes_sha3_ctx_t *ctx);
```
Comment (mguetschow) [inline sys/psa_crypto/include/psa_hashes.h:352]: Would it make sense to have different psa modules for SHA3-256/384/512 or is the increase in code size if including all three per default negligible?

File: tests/sys/psa_crypto_hashes/main.c [hunk 6]
```diff
@@ -41,6 +42,16 @@ int main(void)
         printf("Hash failed: %s\n", psa_status_to_humanly_readable(status));
     }
 
+
+#if IS_USED(MODULE_PSA_RIOT_HASHES_SHA_3)
```
Comment (mguetschow) [inline tests/sys/psa_crypto_hashes/main.c:46]:
```suggestion
```

I would remove that so we could also test an hardware backend with the same code as soon as it is supported.

File: tests/sys/psa_crypto_hashes/main.c [hunk 7]
```diff
@@ -41,6 +42,16 @@ int main(void)
...
+    start = ztimer_now(ZTIMER_USEC);
+    status = sha_3_glue_test();
+    printf("SHA3 glue code test took %d us\n", (int)(ztimer_now(ZTIMER_USEC) - start));
+    if (status != PSA_SUCCESS) {
+        failed = true;
+        printf("SHA3 glue code failed: %s\n", psa_status_to_humanly_readable(status));
+    }
+#endif
```
Comment (mguetschow) [inline tests/sys/psa_crypto_hashes/main.c:54]:
```suggestion
```
matching above

File: tests/sys/psa_crypto_hashes/main.c [hunk 8]
```diff
@@ -23,6 +23,7 @@
 #include "ztimer.h"
 
 extern psa_status_t example_hash(void);
+extern psa_status_t sha_3_glue_test(void);
```
Comment (mguetschow) [inline tests/sys/psa_crypto_hashes/main.c:26]:
```suggestion
extern psa_status_t example_hash_sha3(void);
```

maybe?

File: tests/sys/psa_crypto_hashes/sha3.c [hunk 9]
```diff
@@ -0,0 +1,44 @@
+#include "psa/crypto.h"
+
+static const uint8_t msg[] = "Hello World!";
+static const size_t msg_len = sizeof(msg)-1; // exclude NULL-byte
+
+static const uint8_t hash_sha3_256[] ={
+    0xd0, 0xe4, 0x74, 0x86, 0xbb, 0xf4, 0xc1, 0x6a, 0xca, 0xc2, 0x6f, 0x8b, 0x65, 0x35, 0x92, 0x97, 0x3c, 0x13, 0x62, 0x90, 0x9f, 0x90, 0x26, 0x28, 0x77, 0x08, 0x9f, 0x9c, 0x8a, 0x45, 0x36, 0xaf
```
Comment (mguetschow) [inline tests/sys/psa_crypto_hashes/sha3.c:7]: Please keep the line length below 80 chars, see the RIOT [coding convention](https://github.com/RIOT-OS/RIOT/blob/master/CODING_CONVENTIONS.md) (same below). The CI will complain in any case :)

**Review by mguetschow — CHANGES_REQUESTED** [review]
Nice work! :tada: 

Some rather minor suggestions for improvement below. Please add them first in separate commits and only squash/rebase after we've had a look at the new changes.

File: sys/psa_crypto/include/psa_hashes.h (hunk 5 above)
Comment (Wunderbaeumchen99817) [reply sys/psa_crypto/include/psa_hashes.h:352]: @mguetschow The increase in code-size is indeed negligible since they are all based on the same keccak-code (which is the biggest part of the underlying code), which needs to be compiled anyway with every variant of the sha3 algorithm.

File: sys/hashes/psa_riot_hashes/sha_3.c (hunk 3 above)
Comment (Wunderbaeumchen99817) [reply sys/hashes/psa_riot_hashes/sha_3.c:28]: That's true! Maybe the underlying functions will return something in the future, then the glue-code can also return different error codes :)

File: sys/psa_crypto/include/psa_hashes.h (hunk 5 above)
Comment (mguetschow) [reply sys/psa_crypto/include/psa_hashes.h:352]: Okay, then it sounds reasonable to me to keep it as is. But maybe @Einhornhool has a different opinion and would argue in favor of consistency with the SHA-2 PSA modules?

File: sys/include/hashes/psa/riot_hashes.h (hunk 4 above)
Comment (Wunderbaeumchen99817) [reply sys/include/hashes/psa/riot_hashes.h]: Thanks for the notice! The include is now happening in crypto_includes.h, as for the other hash software implementations too :)

**Review by mguetschow — APPROVED** [review]
Thanks for your fixups, I'm now happy with the changeset! You may rebase and squash everything together in one or two commits (in case you want to split implementation + test in separate commits).

File: sys/psa_crypto/include/psa_hashes.h (hunk 5 above)
Comment (benpicco) [reply sys/psa_crypto/include/psa_hashes.h:352]: > I think separation is only needed once we get a hardware accelerator that only supports a subset of them.

but that would be an API change then

File: sys/psa_crypto/doc.txt [hunk 10]
```diff
@@ -324,6 +324,12 @@
  * - psa_hash_sha_512_custom_backend
  * - psa_hash_sha_512_backend_riot
  *
+ * #### SHA 3
```
Comment (mguetschow) [inline sys/psa_crypto/doc.txt:327]: This is probably the point where you could explicitly mention that this includes SHA3-256, SHA3-384 and SHA3-512.

Also, please move SHA 3 below the truncated SHA512 variants.

File: sys/psa_crypto/include/psa_hashes.h (hunk 5 above)
Comment (Einhornhool) [reply sys/psa_crypto/include/psa_hashes.h:352]: I considered this and I agree. This would be an API change, which makes it more prone to errors.
Also it would require people who add drivers in the future to not only add wrappers, but also to change the module structure.
That makes it less friendly to develop.

I'm sorry for changing my mind now, but I think it would be better to keep the structure consistent with the other modules :)

File: sys/psa_crypto/doc.txt [hunk 11]
```diff
@@ -336,6 +336,12 @@
  * - psa_hash_sha_512_256_custom_backend
  * - psa_hash_sha_512_256_backend_riot
  *
+ * #### SHA 3 - includes SHA3-256, SHA3-384, SHA3-512
+ * - psa_hash_sha_3
+ * - psa_hash_sha_3_backend_periph
+ * - psa_hash_sha_3_custom_backend
+ * - psa_hash_sha_3_backend_riot
+ *
```
Comment (mguetschow) [inline sys/psa_crypto/doc.txt:344]:
```suggestion
 *
```

**Review by mguetschow — CHANGES_REQUESTED** [review]
@Wunderbaeumchen99817 sorry for the back and forth, but could you try to split the three parts in separate (pseudo)modules and add your changes as separate commits for easier review?

**Comment by mguetschow** [comment]
There are still two failing static checks that you should be able to fix (a trailing whitespace and running `make generate-features` once from the main RIOT repo): https://github.com/RIOT-OS/RIOT/actions/runs/9570133114/job/26416056964?pr=20698

I'm not sure why the third check failed, though.

File: features.yaml [hunk 12]
```diff
@@ -827,6 +827,12 @@ groups:
       help: MD5 hardware acceleration present.
     - name: periph_hash_sha_1
       help: SHA-1 hardware acceleration present.
+    - name: periph_hash_sha3_256
+      help: SHA-3/256 hardware acceleration present.
+    - name: periph_hash_sha3_384
+      help: SHA-3/384 hardware acceleration present.
+    - name: periph_hash_sha3_512
+      help: SHA-3/512 hardware acceleration present.
```
Comment (mguetschow) [inline features.yaml:835]: Would you mind moving them below all the SHA-2 functions, here and in the other files? Just for consistency reasons.

File: sys/psa_crypto/Makefile.dep [hunk 13]
```diff
@@ -276,6 +276,78 @@ ifneq (,$(filter psa_hash_sha_512_backend_riot,$(USEMODULE)))
...
+  FEATURES_REQUIRED += periph_hash_sha3_512
+endif
+
+ifneq (,$(filter psa_hash_sha3_512_backend_riot,$(USEMODULE)))
+  USEMODULE += hashes
+  USEMODULE += psa_riot_hashes
+  USEMODULE += psa_riot_hashes_sha3_512
+endif
+
 ## SHA-512/224
```
Comment (mguetschow) [inline sys/psa_crypto/Makefile.dep:351]: Same here and below: Please leave all SHA-2 related parts together.

File: sys/psa_crypto/doc.txt [hunk 14]
```diff
@@ -336,6 +336,23 @@
...
+ * - psa_hash_sha3_256
+ * - psa_hash_sha3_256_backend_periph
+ * - psa_hash_sha3_256_custom_backend
+ * - psa_hash_sha3_256_backend_riot
+ *
+ * #### SHA 3/384
+ * - psa_hash_sha3_384
+ * - psa_hash_sha3_384_backend_periph
+ * - psa_hash_sha3_384_custom_backend
+ * - psa_hash_sha3_384_backend_riot
```
Comment (mguetschow) [inline sys/psa_crypto/doc.txt:349]:
```suggestion
 * - psa_hash_sha3_384_backend_riot
```

File: sys/psa_crypto/doc.txt [hunk 15]
```diff
@@ -336,6 +336,23 @@
...
+ * - psa_hash_sha3_384_backend_periph
+ * - psa_hash_sha3_384_custom_backend
+ * - psa_hash_sha3_384_backend_riot 
+ *
+ * #### SHA 3/512
+ * - psa_hash_sha3_512
+ * - psa_hash_sha3_512_backend_periph
+ * - psa_hash_sha3_512_custom_backend
+ * - psa_hash_sha3_512_backend_riot
  * ### MAC
```
Comment (mguetschow) [inline sys/psa_crypto/doc.txt:356]:
```suggestion
 * - psa_hash_sha3_512_backend_riot
 *
 * ### MAC
```

File: sys/psa_crypto/include/psa_hashes.h [hunk 16]
```diff
@@ -341,6 +341,109 @@ psa_status_t psa_hashes_sha512_256_finish(psa_hashes_sha512_256_ctx_t *ctx,
...
+ * @param ctx
+ * @param hash
+ * @return psa_status_t
+ */
+psa_status_t psa_hashes_sha3_256_finish(psa_hashes_sha3_ctx_t *ctx,
+                                      uint8_t *hash);
+#endif /* MODULE_PSA_HASH_SHA3_256 */
+
+#if IS_USED(MODULE_PSA_HASH_SHA3_384) || defined(DOXYGEN)/**
+ * @brief   Low level wrapper function to call a driver for a general SHA3 hash update
```
Comment (mguetschow) [inline sys/psa_crypto/include/psa_hashes.h:380]:
```suggestion
#if IS_USED(MODULE_PSA_HASH_SHA3_384) || defined(DOXYGEN)
/**
 * @brief   Low level wrapper function to call a driver for a general SHA3 hash update
```

File: sys/psa_crypto/include/psa_hashes.h [hunk 17]
```diff
@@ -341,6 +341,109 @@ psa_status_t psa_hashes_sha512_256_finish(psa_hashes_sha512_256_ctx_t *ctx,
...
+ * @param ctx
+ * @param hash
+ * @return psa_status_t
+ */
+psa_status_t psa_hashes_sha3_384_finish(psa_hashes_sha3_ctx_t *ctx,
+                                      uint8_t *hash);
+#endif /* MODULE_PSA_HASH_SHA3_384 */
+
+#if IS_USED(MODULE_PSA_HASH_SHA3_512) || defined(DOXYGEN)/**
+ * @brief   Low level wrapper function to call a driver for a general SHA3 hash update
```
Comment (mguetschow) [inline sys/psa_crypto/include/psa_hashes.h:414]:
```suggestion
#if IS_USED(MODULE_PSA_HASH_SHA3_512) || defined(DOXYGEN)
/**
 * @brief   Low level wrapper function to call a driver for a general SHA3 hash update
```

File: sys/psa_crypto/psa_crypto_algorithm_dispatch.c [hunk 18]
```diff
@@ -152,6 +176,18 @@ psa_status_t psa_algorithm_dispatch_hash_update(psa_hash_operation_t *operation,
...
+    #if (IS_USED(MODULE_PSA_HASH_SHA3_384))
+    case PSA_ALG_SHA3_384:
+        return psa_hashes_sha3_384_update(&operation->ctx.sha3, input, input_length);
+    #endif
+    #if (IS_USED(MODULE_PSA_HASH_SHA3_512))
+    case PSA_ALG_SHA3_512:
+        return psa_hashes_sha3_512_update(&operation->ctx.sha3, input, input_length);
+    #endif
     #if (IS_USED(MODULE_PSA_HASH_SHA_512_224))
     case PSA_ALG_SHA_512_224:
```
Comment (mguetschow) [inline sys/psa_crypto/psa_crypto_algorithm_dispatch.c:192]: ... and here (SHA-3 after SHA-2) :P

**Review by mguetschow — CHANGES_REQUESTED** [review]
Looks mostly good, just some more minor nitpicks :)

File: sys/psa_crypto/Makefile.dep [hunk 19]
```diff
@@ -276,6 +276,64 @@ ifneq (,$(filter psa_hash_sha_512_backend_riot,$(USEMODULE)))
...
 
+ifneq (,$(filter psa_hash_sha3_512_backend_periph,$(USEMODULE)))
+  FEATURES_REQUIRED += periph_hash_sha3_512
+endif
+
+ifneq (,$(filter psa_hash_sha3_512_backend_riot,$(USEMODULE)))
+  USEMODULE += hashes
+  USEMODULE += psa_riot_hashes
+  USEMODULE += psa_riot_hashes_sha3_512
+endif
```
Comment (mguetschow) [inline sys/psa_crypto/Makefile.dep:287]: It seems you missed to move those down.

Comment (darthdrannel) [reply sys/psa_crypto/Makefile.dep]: good catch, has been done now

**Review by mguetschow — APPROVED** [review]
Looks good to me now, please squash @Wunderbaeumchen99817 :)

**Comment by Wunderbaeumchen99817** [comment]
@mguetschow done :)

**Review by mguetschow — APPROVED** [review]
Thanks, nice work!
